Improve shell stream

294797392 opened the pull request:
Improve ShellStream

* Added SendWindowChangeRequest to ShellStream

* Added a new constructor to ShellStream

* SshClient adds an overload function for CreateShellStream

When creating ShellStream, the channel will be immediately opened in the constructor of ShellStream. At this time, the calling module may not have subscribed to ShellStream's DataReceived event, which may result in the loss of a small piece of data immediately after connection.
You can use ShellStream's new constructor to solve this problem

294797392 committed:
Improve ShellStream

* Added SendWindowChangeRequest to ShellStream

* Added a new constructor to ShellStream

When creating ShellStream, the channel will be immediately opened in the constructor of ShellStream. At this time, the calling module may not have subscribed to ShellStream's DataReceived event, which may result in the loss of a small piece of data immediately after connection.
You can use ShellStream's new constructor to solve this problem

294797392 committed:
Improve ShellStream

* Added SendWindowChangeRequest to ShellStream

* Added a new constructor to ShellStream

* SshClient adds an overload function for CreateShellStream

When creating ShellStream, the channel will be immediately opened in the constructor of ShellStream. At this time, the calling module may not have subscribed to ShellStream's DataReceived event, which may result in the loss of a small piece of data immediately after connection.
You can use ShellStream's new constructor to solve this problem

294797392 commented:
Why does AppVeyor fail?

294797392 commented:
Why does appveyor fail?

294797392 committed:
Merge branch 'develop' into ImproveShellStream

Rob-Hague commented on src/Renci.SshNet/ShellStream.cs:
```diff
@@ -83,6 +83,29 @@ private void AssertValid()
+        /// <exception cref="SshException">The pseudo-terminal request was not accepted by the server.</exception>
+        /// <exception cref="SshException">The request to start a shell was not accepted by the server.</exception>
+        internal ShellStream(ISession session, string terminalName, uint columns, uint rows, uint width, uint height, IDictionary<TerminalModes, uint> terminalModeValues, int bufferSize, EventHandler<ShellDataEventArgs> dataReceived, EventHandler<ExceptionEventArgs> errorOccurred)
+               : this(session, terminalName, columns, rows, width, height, terminalModeValues, bufferSize)
+        {
+            DataReceived += dataReceived;
+            ErrorOccurred += errorOccurred;
```
Since the other constructor will run before these two lines, how does this solve your problem?

294797392 replied:
Sorry, there was a slight issue with the modifications I made

Rob-Hague commented:
Hi @294797392 

Are you able solve your problem using the `Read` method instead of the `DataReceived` event? e.g.

```c#
byte[] buffer = new byte[1024];
int bytesRead;
while ((bytesRead = shellStream.Read(buffer) != 0)
{
    // Process "bytesRead" bytes from buffer
}

```

294797392 commented:
Yes, using the Read function can solve my problem.
But I think it's best to optimize it, as other developers may also encounter this issue when using it.
Do you think it is necessary to add the SendWindowChangeRequest function?
@Rob-Hague

Rob-Hague commented:
Yes it seems reasonable to add SendWindowChangeRequest - there have been previous requests for it. Personally I would suggest to:

1) make it return `void`: the server does not respond to a window-change request, and our implementation in `ChannelSession` just returns true so it is not much use to return `bool`
2) add `ThrowHelper.ThrowObjectDisposedIf(_disposed, this);` at the start of the method
3) add at least a basic test. For example, in `test\Renci.SshNet.Tests\Classes\ShellStreamTest.cs` there could be a test which verifies the call to `IChannelSession` mock, and verifies the `ObjectDisposedException` when the ShellStream is closed

Let me know what you think

294797392 commented:
I agree with your idea. 
I tried to modify the code and added the SendWindowChangeRequest method, but found that if it returns void, IDE0058 compilation error will occur.
If void is returned, there may be many modifications and it will also affect the consistency of the code (there are many methods in ChannelSession that directly return true). Perhaps having a return value would be better than having no return value, as a return value can prompt the caller about the execution result of the method you called, and in the future we can also add more error handling to the method (if necessary)!
And I believe that when developing the ChannelSession module initially, there must have been a reason for returning a boolean.
@Rob-Hague

Rob-Hague commented:
Ok, bool seems fine to me - I don't feel strongly about it

294797392 commented:
> Ok, bool seems fine to me - I don't feel strongly about it

Do you agree with me changing it to a boolean return value? If you agree, I will submit another PR

Rob-Hague commented:
Yes

294797392 committed:
1. Add a check on whether ChannelSession is disposed when calling SendWindowChangeRequest

2. Added SendWindowChangeRequest_ThrowsObjectDisposedException to ShellStreamTest

Rob-Hague commented on test/Renci.SshNet.Tests/Classes/ShellStreamTest.cs:
```diff
@@ -123,6 +123,18 @@ public void Write_AfterDispose_ThrowsObjectDisposedException()
         }
 
+        [TestMethod]
+        public void SendWindowChangeRequest_ThrowsObjectDisposedException()
```
Could you also add a simple test which verifies the call to `ChannelSession.SendWindowChangeRequest`?

Rob-Hague reviewed (commented):
Could you please change `ChannelSession.SendWindowChangeRequest` to call `TrySendMessage` instead of `SendMessage`?
